Misc. intel_adsp cleanup (big patch, no logic changes)

andyross opened the pull request:
Now that features are locked down I finally have time to finish and push out some of the cleanup that keeps getting delayed.  No rush on this PR, it's not a bug fix and strictly shouldn't change logic at all (beyond producing slightly smaller binaries).

See the various patches for details.  Most of this consists of taking code that was duplicated across platforms and unifying it, to the tune of ~1500 lines of shrinkage.

It's not done, though this should be reviewable in isolation.  I'm working on the linker script (and memory.h header) right now to produce a single variant that is parametrized only by devicetree and the core-isa.h.

Commit:
soc: intel_adsp: Unify IDC drivers, eliminate IPM_CAVS_IDC

Have all platforms use the same underling IPI mechanism.
Unfortunately I can't remove the ipm_cavs_idc driver entirely as it's
still used by the intel_s1000 board (which also should be unified with
cavs, but I don't have that hardware anymore).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: General soc_mp.c cleanup

Misc cleanup, no non-trivial logic changes.

Swap in new ("rsr <reg>, REGISTER_NAME") syntax for Xtensa SR's in
place of inconsistent usage of the older one ("rsr.REGISTER_NAME
<reg>").

Remove the legacy handling of !KERNEL_COHERENCE cases for allocating
the cpu start record.  That has long been a requirement of
multiprocessor code on this platform.

Remove the synchronous testing of the "alive" flag in
arch_start_cpu().  Nothign about that API is intended to be
synchronous, and in fact the Zephyr SMP layer is already doing the
same trick.

Remove some vestigial dead code at the end of z_mp_entry().  It was
apparently intended to handle the case where a CPU function returned,
but that's not legal anyway.  And it was only enabled in the case
where there was only one CPU anyway, which was an impossible situation
(you can't evercall arch_start_cpu() successfully on a system with
only one core, for obvious reasons -- the only core is already
running!).  Replace with an assertion.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers/ipm: Remove intel_adsp_mailbox driver

This is dead code.  It's based on the cAVS "IPC" mechanism to allow
communication to and from the host CPU.  But there is no test rig in
the Zephyr tree for the protocol defined.  And in fact the only
Zephyr-based user of the IPC mechanism (Sound Open Firmware) has its
own IPC driver and speaks its own protocol with the host kernel.  That
driver needs to migrate into Zephyr soon and this legacy bit is just
confusing.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_asdp: Clean up soc_init() code

Reorganize the initialization code to cleanly separate the platforms
and clarify which code is common.  The #if'ery was sort of a mess.
This is in preparation for an incoming patch that unifies the shim
register definitions across platform variants.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Remove old support for cAVS 2.5 "alternate boot"

The linker script (and a little bit of SOF) still has support for an
older mechanism for bootstrapping secondary cores by copying code into
lp-sram from a "manifest" emitted by the linker.  This actually never
worked in Zephyr, and we've implemented a different scheme that uses a
small runtime-copied trampoline instead.

Remove.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Clean up shim driver

Each platform was defining its own shim.h header, with slightly
variant field definitions, for a register block that is almost
completely compatible between versions.  This is made worse by the
fact that these represent an API imported fairly early from SOF, the
upstream version of which has since diverged.

Move the existing shim struct into a header ("cavs-shim.h") of its
own, remove a bunch of unused symbols, fill in definitions for some
registers that were left out, correct naming to match the hardware
docs in a few places, make sure all hardware dependencies are source
from devicetree only, and modify existing usage to use the new API
exclusively.

Interestingly this leaves the older shim.h header in place, as it
turns out to contain definitions for a bunch of things that were never
part of the shim register block.  Those will be unified in separate
patches.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
soc: intel_adsp: Unify bootloader configuration

The generic bootloader code used a per-device "platform.h" file
imported from SOF.  These turn out to have very little actual content.
Move them to the core directory in a single header for now, pending
some rework to place the settings in devicetree.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

lyakh reviewed (changes requested):
I see several logic changes in this PR and I've just tested to confirm that it breaks SOF on APL multicore

andyross commented:
@lyakh, can I get you to elaborate?  From my testing, it's working in the Zephyr test suite and initializes two cores correctly on SOF with the rig I had last week.  Where is it failing and do you have an exerciser?  As mentioned, I'd really, really like to get to a point where I can run the entire SOF test suite, but it's complicated by our use of different hardware.

As far as "logic changes", there really aren't any, I promise.  Nothing significant anyway -- it's all just refactoring to eliminate duplication.

Again, this is just here for early review.  There's no need to rush this in for the release.

lyakh commented:
@andyross my test case is still the same and very simple: try to boot an SMP topology on APL. The "boot core 1" IPC succeeded but then the first IPC that had to be executed on core 1 failed. Switching back to the current Zephyr "main" fixed the problem.
As for logic changes - at least I wasn't able to verify that the logic before and after the change is the same. Let me add some comments in the code where that was the case. Not in all of them, just a couple of examples.

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -251,9 +250,24 @@ void z_mp_entry(void)
+	if(IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V15)) {
+		k_busy_wait(10);
+	}
+	IDC[start_rec.cpu].busy_int = IDC_ALL_CORES;
```
isn't this a logic change?

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -354,27 +368,18 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 
 	IDC[curr_cpu].core[cpu_num].ietc = ietc;
-	IDC[curr_cpu].core[cpu_num].itc = IDC_MSG_POWER_UP | IPC_IDCITC_BUSY;
+	IDC[curr_cpu].core[cpu_num].itc = IDC_MSG_POWER_UP | BIT(31);
```
not a logic change, but why? Even if it's the same value on all platforms, "BUSY" tells me much more than 31.

andyross replied:
Looks like a mis-squash, will fix.

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -354,27 +368,18 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
-	cavs_idc_smp_init(NULL);
-#endif
+	k_busy_wait(10);
+	IDC[cpu_num].busy_int &= ~IDC_ALL_CORES;
```
logic change?

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -383,25 +388,13 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
-
-	ipm_send(idcdev, 0, IPM_CAVS_IDC_MSG_SCHED_IPI_ID,
-		 IPM_CAVS_IDC_MSG_SCHED_IPI_DATA, 0);
-#endif
```
here too?

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -218,17 +193,17 @@ void z_mp_entry(void)
-#ifndef CONFIG_SOC_SERIES_INTEL_CAVS_V25
-	z_xtensa_cache_flush_inv_all();
-#endif
+	if (!IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V25)) {
```
is there any rule / recommendation when to use preprocessor and when (pseudo) run-time?

andyross replied:
Pretty much kernel style: use IS_ENABLED() everywhere you can because it's more readable.  Use #if's only where needed because the included code won't always be able to build successfully without the needed configuration (and even then we try to make sure that doesn't happen by e.g. not having preprocessor-dependent symbols defined, etc....)

lyakh commented on soc/xtensa/intel_adsp/common/soc_mp.c:
```diff
@@ -381,9 +344,6 @@ void arch_start_cpu(int cpu_num, k_thread_stack_t *stack, int sz,
 #endif
-
-	while (!start_rec.alive)
-		;
```
wouldn't this let the primary core return early without waiting for the secondary core to boot and let the host start using the secondary core before it's ready?

andyross replied:
No, because that logic is already being done in the portable SMP layer in the Zephyr kernel.  arch_start_cpu() is not required to be atomic, this was just wasted bytes.

andyross commented:
Huh.... This was on an Up Squared board?  Which topology file are you using, and built from which SOF version?  I swear I checked this through intialiization on both APL and TGL and it was working for me.

As far as the rest of the comments... not really understanding the nitpicking about the semantics of "logic change".  I'm not promising zero changes to the output binary.  Some code is getting moved, some reordered, some is being accessed through different APIs that will generate different code, etc...  The point to the promise is to streamline review by promising that no behaviors should be changing and that if they do it's a clear bug.

Would you be OK if I retitled the PR to "no significant logic changes"?

lyakh commented:
> Huh.... This was on an Up Squared board? Which topology file are you using, and built from which SOF version? I swear I checked this through intialiization on both APL and TGL and it was working for me.

This topology isn't upstream, I just took the upstream sof-apl-pcm512x topology and moved one of the HDMI pipelines to core 1. But I'm not sure you can test that - do you have a HiFiBerry card with your UP2? If not, are you using the nocodec or the hda-generic topology? I'm attaching the modified topology in case you have a HiFiBerry.
[sof-apl-pcm512x-smp.tplg.zip](https://github.com/zephyrproject-rtos/zephyr/files/7116824/sof-apl-pcm512x-smp.tplg.zip)

> As far as the rest of the comments... not really understanding the nitpicking about the semantics of "logic change". I'm not promising zero changes to the output binary. Some code is getting moved, some reordered, some is being accessed through different APIs that will generate different code, etc... The point to the promise is to streamline review by promising that no behaviors should be changing and that if they do it's a clear bug.

Sorry, I was trying to actually follow in your patches how the logic is preserved and I couldn't. It isn't about the naming of course. If you confirm, that the cases that I mentioned are indeed preserving logic, I'll have to dig deeper into them.

> Would you be OK if I retitled the PR to "no significant logic changes"?

andyross commented:
Are you worried about the removal of the IPM driver maybe?  Indeed, OK, that's a "logic change" when viewed from the generated code of 1.5.  It's now using the same code and same driver as cAVS 2.5.  I guess I view that as "unification" since neither code path significantly changed, but OK, you got me.  That's a logic change. :)

Commit:
tests/kernel: Add test for CONFIG_SMP_BOOT_DELAY

This feature is in tree and used by the SOF app, but we don't have a
local test for it.  Add one, including a case to track regressions in
a known failure mode (where the second CPU wouldn't get its IDC
interrupts set up correctly if spawned at runtime).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
drivers: cavs_timer: Use the new interrupt controller API

Recent work to this platform added a new, cleaner low level API to the
interrupt controller.  Replace the hand-cooked register access with
that.  This is still not as good as having proper multicore support in
the intc_cavs driver, but it's at least better.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Correct some checkpatch issues, add a last patch with a test of SMP_BOOT_DELAY/z_smp_start_cpu(), and split out a build fix for that (also need SOF PR https://github.com/thesofproject/sof/pull/4726) into a separate patch also submtted in #38368

kv2019i commented on soc/xtensa/intel_adsp/common/bootloader/boot_entry.S:
```diff
@@ -14,10 +14,12 @@
 #include <xtensa/config/core-isa.h>
+#include "platform.h"
+
+#define PLATFORM_PRIMARY_CORE_ID 0
```
I don't know how this happens, but with this PR, 

```
sof/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/bootloader/boot_entry.S: Assembler messages:
sof/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/bootloader/boot_entry.S:58: Error: missing ')'
sof/zephyrproject/zephyr/soc/xtensa/intel_adsp/common/bootloader/boot_entry.S:58: Error: junk at end of line, first unrecognized character is `U'
xt-xcc ERROR:  /xtensa/XtDevTools/install/tools/RG-2017.8-linux/XtensaTools/bin/xt-as returned non-zero status 1
```

Looking at preprocessor output for boot_entry.S, it seems the xcc-as doesn't like Zephyr BIT(1) definition in assembly files.

```
     58 wnd0_base_val:
     59  .word (0xbe000000 + 0x4000) | (1UL << (1)) | (1UL << (0))
```

andyross replied:
Heh, nice synchronicity.  I bumped into exactly this error (though possibly in a different place) building the Zephyr thread_apis test with xcc testing the (unrelated) linker fix yesterday.  I haven't gotten back to investigate.  But yeah, apparently there's something glitchy here.  Maybe the assembler doesn't like ANSI "UL" suffixes?  Surely it supports bit shifting...

kv2019i replied:
Works now in lates version.

kv2019i reviewed (changes requested):
Logging a weird build fail with XCC, see inline.

Commit:
soc/intel_adsp: Fix _ASMLANGUAGE declaration for boot_entry.S

The cAVS bootloader code (for... somewhat silly reasons) doesn't build via
the same paths as regular Zephyr object files, so it wasn't getting
the _ASMLANGUAGE define.  That meant that Zephyr headers defining
BIT() were using syntax incompatible with some assemblers (specifically
the Cadence xcc assembly; current gas versions were fine).

Not 100% sure this is the best spot to put this, but the root fix is
to get the bootloader building into the same link as the rest of Zephyr
anyway.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Rebased.  Includes a fix for the boot_entry.S build failure (which was always a bug, but only unmasked here -- prior to this series that file was seeing a SOF-defined BIT() macro and not Zephyr's).

kv2019i reviewed (approved):
Looks good now. The IDC refactoring patch is stretching the "no logic changes" headline a bit and might benefit from more review scrutiny. I tested build with TGL and APL and run the multicore tplg test on TGL and still working. Need @lyakh  to test on APL multicore. There's also a few comments from @lyakh open (mostly in the IDC refactoring patch).

marc-hb commented:
The ability to submit Zephyr PRs like this one to SOF CI should come very soon, please help review the first part:
https://github.com/thesofproject/sof/pull/4740
Digression over sorry.

marc-hb commented:
> The ability to submit Zephyr PRs like this one to SOF CI should come very soon,

It works  now, first demo at https://github.com/thesofproject/sof/pull/4761

Requires a one-line change in one SOF script + a SOF PR. It's like updating a west module but in reverse https://github.com/zephyrproject-rtos/zephyr/discussions/34713
